oxidation/libparsec/crates/client-connection/src/authenticated_cmds.rs

```diff
@@ -0,0 +1,318 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+
+//! Send Authenticated commands to the server.
+//!
+//! The HTTP-request will contain the following headers:
+//!
+//! ```text
+//! Authorization: PARSEC-SIGN-ED25519
+//! Author: [base64 string]
+//! Timestamp: [timestamp in millisecond]
+//! Signature: [base64 ed25519 signature]
+//! ```
+//! The signature is generated by appending the following data:
+//!
+//! 1. `author` (in base64)
+//! 2. `timestamp` (u128 in big-endian)
+//! 3. `body` (as bytes)
+
+use std::{collections::HashMap, num::NonZeroU64};
+
+use parsec_api_crypto::{PublicKey, SigningKey};
+use parsec_api_protocol::authenticated_cmds::{
+    self, invite_delete::InvitationDeletedReason, invite_new::UserOrDevice,
+};
+use parsec_api_types::{
+    BlockID, DateTime, InvitationToken, RealmID, ReencryptionBatchEntry, UserID, VlobID,
+};
+use reqwest::{
+    header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_LENGTH, CONTENT_TYPE},
+    Client, RequestBuilder, Url,
+};
+
+use crate::command_error::{self, CommandError};
+
+/// Endpoint path were will be sending authenticated cmds.
+pub const AUTHENTICATED_API_URI: &str = "/authenticated";
+/// Method name that will be used for the header `Authorization` to indicate that will be using this method.
+pub const PARSEC_AUTH_METHOD: &str = "PARSEC-SIGN-ED25519";
+/// How we serialize the data before sending a request.
+pub const PARSEC_CONTENT_TYPE: &str = "application/msgpack";
+
+/// Factory that send commands in a authenticated context.
+pub struct AuthenticatedCmds {
+    /// HTTP Client that contain the basic configuration to communicate with the server.
+    client: Client,
+    url: Url,
+    user_id: String,
+    signing_key: SigningKey,
+}
+
+impl AuthenticatedCmds {
+    /// Create a new `AuthenticatedCmds`
+    pub fn new(
+        client: Client,
+        root_url: &str,
```

Comment:
I cannot simply, change the type because of the test.

I change the signature to use `root_url: AsRef<str>`

Reply:
why can't you change the test as well ?

Reply:
I don't find `BackendOrganizationAddr` easy to use

Reply:
> I don't find BackendOrganizationAddr easy to use

`BackendOrganizationAddr` implements FromStr so it should only add an `.unwrap()` to your string in the test, do you see something that could be improved there ?

Having a strong typing on our different kind of url is import given it's otherwise easy to get lost and use the wrong one, on top of that we end up needing to do additional error checks (hence the need for a Result here) which makes things more complex for no gain

Reply:
btw you may have been puzzled by the fact `BackendOrganizationAddr`'s FromStr accept url with parsec:// scheme, this is because we don't want to have this address directly opened by a webbrowser when displayed to a user (but there is also a redirection representation of this kind of url which this time have a http:// scheme... so in the end handling url is complicated and hence it's simpler to let `BackendOrganizationAddr` handle this ^^)

Reply:
I didn't see the impl `FromStr` that was filled by `impl_common_stuff`